core/api/src/main/scala/org/virtuslab/ideprobe/jsonrpc/logging/RequestResponseLogger.scala

```diff
@@ -0,0 +1,93 @@
+package org.virtuslab.ideprobe.jsonrpc.logging
+
+import java.time.Instant
+import java.util.concurrent.{ScheduledFuture, ScheduledThreadPoolExecutor}
+import scala.annotation.tailrec
+import scala.concurrent.duration.{Duration, SECONDS}
+
+class RequestResponseLogger {
+
+  private val ex = new ScheduledThreadPoolExecutor(1)
+  private val flushTask: Runnable = () => flush()
+  private var scheduledTask: Option[ScheduledFuture[_]] = None
+
+  private val flushTimeout = Duration(1, SECONDS)
+
+  private sealed trait LogEntry
+  private case class Request(message: String) extends LogEntry
+  private case class Response(message: String) extends LogEntry
+  private case class RequestAndResponse(request: String, response: String) extends LogEntry
+
+  private var buffer: List[(LogEntry, Instant)] = List.empty
+
+  def logRequest(request: String): Unit = buffer.synchronized {
```

Comment:
Synchronization is rather not needed at all, but if you synchronize on something, it would make more sense to synchronize on `this`. The buffer is changing.

Reply:
Kept the synchronization to avoid potential clashes with scheduled flushing. Now syncing on `this`.